File: src/baekjoon/step/graph/PROB24479.java
package baekjoon.step.graph;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.StringTokenizer;

public class PROB24479 {
    /**
     * 오름차순으로 인접 정점 방문
     * i 번 째 줄에는 정점 i 의 방문 순서를 출력한다.
     * 방문 할 수 없는 경우 0 을 출력한다.
     */

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringBuilder sb = new StringBuilder();
    static StringTokenizer st;
    static int N, M, R;
    static ArrayList<ArrayList<Integer>> graph = new ArrayList<>(); // 그래프
    static boolean[] visited; // 방문 여부 판단
    static int[] result; // 순번 저장
    static int cnt = 1;

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());

        visited = new boolean[N + 1];
        result = new int[N + 1];

        for (int i = 0; i <= N; i ++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < M; i ++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken()); // 정점
            int v = Integer.parseInt(st.nextToken()); // 정점

            // 양방향 간선
            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        for (int i = 1; i <= N; i ++) {
            Collections.sort(graph.get(i));
        }

        dfs(R);

        for (int i = 1; i <= N; i ++) {
            sb.append(result[i]).append("\n");
        }
        System.out.println(sb);
    }

    public static void dfs(int currentNode) {
        visited[currentNode] = true;
        result[currentNode] = cnt ++;

        for (Integer value : graph.get(currentNode)) {
            if (!visited[value]) {
                dfs(value);
            }
        }
    }
}

File: src/baekjoon/step/graph/PROB24479.py
# 첫번째 줄 N : 정점의 수  M : 간선의 수 R : 시작 정점
# 두번째 줄 u : 정점 v : 가중치 1 인 양방향 간선


import sys
# 재귀 에러 방지
sys.setrecursionlimit(10**6)


N, M, R = map(int, input().split())

# 그래프 생성
# 1 번 노드와 인덱스 값이 같게 하기 위해서 N + 1 로 둔다.
graph = [[] for i in range(N + 1)]

for i in range(M) :
    u, v = map(int, sys.stdin.readline().split())
    graph[u].append(v)
    graph[v].append(u)

# 방문 리스트
visited = [0] * (N + 1)

# 방문 순서
visited_order = 1

# DFS 함수
def dfs(v) :
    global visited_order
    visited[v] = visited_order
    # 정점을 오름차순으로 방문한다고 했으므로, 그래프를 정렬 해준다.
    graph[v].sort()

    for i in graph[v] :
        # 방문한 노드가 이전에 방문한 곳이 아니라면 순서를 올려준다.
        if visited[i] == 0 :
            visited_order += 1
            dfs(i)

# 함수 호출
dfs(R)

# 방문 순서 출력
for i in range(1, N + 1) :
    print(visited[i])